File: ruby/day_2.rb
# frozen_string_literal: true

require 'csv'

def repeating_pattern?(number)
  str = number.to_s
  return false if str.length.odd?

  str[0, str.length / 2] == str[str.length / 2..]
end

def part_one(ranges)
  total = ranges.sum do |range|
    min, max = range.split('-').map(&:to_i)
    (min..max).sum { |num| repeating_pattern?(num) ? num : 0 }
  end

  puts "Total sum of all invalid combinations: #{total}"
end

part_one(CSV.read('product_lines.csv').first)

def repeated_sequence?(number) # rubocop:disable Metrics/MethodLength
  str = number.to_s
  length = str.length

  # Try all possible pattern lengths from 1 to length/2
  (1..length / 2).each do |pattern_length|
    # Check if the string length is divisible by pattern length
    next unless (length % pattern_length).zero?

    # Check if every character matches the pattern without creating new strings
    is_repeating = true
    (0...length).each do |i|
      if str[i] != str[i % pattern_length]
        is_repeating = false
        break
      end
    end

    return true if is_repeating
  end

  false
end

def part_two(ranges)
  total = ranges.sum do |range|
    min, max = range.split('-').map(&:to_i)
    (min..max).sum { |num| repeated_sequence?(num) ? num : 0 }
  end

  puts "Part Two - Total sum of all invalid combinations: #{total}"
end

part_two(CSV.read('product_lines.csv').first)

File: python/day_2.py
import csv
from pathlib import Path


def repeating_pattern(number: int) -> bool:
    """Check if number is made of two identical halves (part one logic)."""
    s = str(number)
    if len(s) % 2 != 0:
        return False
    mid = len(s) // 2
    return s[:mid] == s[mid:]


def repeated_sequence(number: int) -> bool:
    """Check if number is made of any pattern repeated at least twice."""
    s = str(number)
    length = len(s)
    
    for pattern_length in range(1, length // 2 + 1):
        if length % pattern_length != 0:
            continue
        
        is_match = True
        for i in range(pattern_length, length):
            if s[i] != s[i % pattern_length]:
                is_match = False
                break
        
        if is_match:
            return True
    
    return False


def part_one(ranges: list[str]) -> int:
    """Sum all numbers that match the repeating pattern."""
    total = sum(
        num
        for range_str in ranges
        for start, end in [range_str.split('-')]
        for num in range(int(start), int(end) + 1)
        if repeating_pattern(num)
    )
    print(f"Part One - Total sum of invalid combinations: {total}")
    return total


def part_two(ranges: list[str]) -> int:
    """Sum all numbers that are made of repeated sequences."""
    total = sum(
        num
        for range_str in ranges
        for start, end in [range_str.split('-')]
        for num in range(int(start), int(end) + 1)
        if repeated_sequence(num)
    )
    print(f"Part Two - Total sum of invalid combinations: {total}")
    return total


if __name__ == "__main__":
    input_path = Path(__file__).parent.parent / "product_lines.csv"
    
    with input_path.open() as f:
        ranges = next(csv.reader(f))
    
    part_one(ranges)
    part_two(ranges)
